### PR TITLE
feat: UseEnumTagCompleter (related to #6076)

### PR DESCRIPTION
**Pull request by xdesou01**
Requires some debugging, but almost finished.

Update: Finished and ready for review. Screenshots below

image
image

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Ready for review/merge?

**Comment by magnus-madsen** [comment]
!!!jar

**Comment by xdesou01** [comment]
> Ready for review/merge?

Yes

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/Completion.scala [hunk 1]
```diff
@@ -222,6 +222,14 @@ sealed trait Completion {
         documentation = None,
         kind = CompletionItemKind.Method
       )
+    case Completion.UseEnumTagCompletion(sym, cas) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/completion/Completion.scala:225]: We write `caze` to avoid the clash with `case`.

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/Completion.scala [hunk 2]
```diff
@@ -467,6 +475,13 @@ object Completion {
    */
   case class UseDefCompletion(name: String) extends Completion
 
+  /**
+   * Represents a Use Enum Tag completion
+   *
+   * @param enumSym the sym of the enum.
+   * @param cas     the case of the enum.
+   */
+  case class UseEnumTagCompletion(enumSym: EnumSym, cas: TypedAst.Case) extends Completion
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/completion/Completion.scala:484]: -> caze

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/completion/Completion.scala:484]: enumSym: EnumSym -> sym: Symbol.EnumSym

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/Completion.scala [hunk 3]
```diff
@@ -467,6 +475,13 @@ object Completion {
    */
   case class UseDefCompletion(name: String) extends Completion
 
+  /**
+   * Represents a Use Enum Tag completion
+   *
+   * @param enumSym the sym of the enum.
+   * @param cas     the case of the enum.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/completion/Completion.scala:482]: caze

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/UseEnumTagCompleter.scala [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  }
+  private def stripWord(ctx: CompletionContext): Option[String] = {
+    val regex = raw"\s*use\s+(.*)".r
+    ctx.prefix match {
+      case regex(word) => Some(word)
+      case _ => None
+    }
+  }
+
+  private def matches(cSym: CaseSym, tag: String): Boolean = cSym.name.startsWith(tag)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/completion/UseEnumTagCompleter.scala:80]: cSym: CaseSym -> sym: Symbol.CaseSym

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/UseEnumTagCompleter.scala [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+ */
+
+package ca.uwaterloo.flix.api.lsp.provider.completion
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.api.lsp.Index
+import ca.uwaterloo.flix.language.ast.TypedAst
+import ca.uwaterloo.flix.language.ast.Symbol
+import ca.uwaterloo.flix.api.lsp.provider.completion.Completion.UseEnumTagCompletion
+import ca.uwaterloo.flix.language.ast.Symbol.{EnumSym, CaseSym, mkEnumSym}
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/completion/UseEnumTagCompleter.scala:24]: Drop these imports, we always prefix with Symbol

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/UseEnumTagCompleter.scala [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+import ca.uwaterloo.flix.language.ast.{SourceLocation, Type, TypeConstructor, TypedAst}
+
+
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/completion/UseEnumTagCompleter.scala:27]: extra line break?

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/UseEnumTagCompleter.scala [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+              Some(UseEnumTagCompletion(sym, cas))
+            }
+            else {
+              None
+            }
+        }
+      case None => Nil
+    }
+  }
+  private def stripWord(ctx: CompletionContext): Option[String] = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/completion/UseEnumTagCompleter.scala:72]: missing line break?

**Comment by magnus-madsen** [comment]
I also find the logic a bit hard to follow. Could you try to add some comments?

**Comment by magnus-madsen** [comment]
I am also not sure about the creation of EnumSym. Usually we don't do that.

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/Completion.scala (hunk 1 above)
Comment (xdesou01) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/completion/Completion.scala:225]: Ah, I've just seen it as `cas` elsewhere but I'll change it

**Comment by xdesou01** [comment]
> I am also not sure about the creation of EnumSym. Usually we don't do that.

That was done in EnumTagCompleter. There's a justification there, but maybe I'll include it here as well. The example is that the respective namespace and name of `A.B.C.Color` would become `A.`, `B.C.Color` from Symbol.mkEnumSym.

**Comment by magnus-madsen** [comment]
> > I am also not sure about the creation of EnumSym. Usually we don't do that.
> 
> That was done in EnumTagCompleter. There's a justification there, but maybe I'll include it here as well. The example is that the respective namespace and name of `A.B.C.Color` would become `A.`, `B.C.Color` from Symbol.mkEnumSym.

Hmm okay.

**Comment by magnus-madsen** [comment]
But in any case could you add some more comments, maybe with an example like done in EnumTagCompleter.

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
Thanks!
